metering/llms.py

```
@@ -0,0 +1,241 @@
+import atexit
+import logging
+import os
+from time import time
+from metering.ingest import create_ingest_client
+from metering.ingest.api_client import create_ingest_payload
+
+
+INPUT_TOKENS_METER_API_NAME = "input_tokens"
+OUTPUT_TOKENS_METER_API_NAME = "output_tokens"
+ANTHROPIC_PROVIDER = "anthropic"
+OPENAI_PROVIDER = "openai"
+COHERE_PROVIDER = "cohere"
+VERTEXAI_PROVIDER = "vertexai"
+UNKNOWN_PROVIDER = "unknown"
+
+logger = logging.getLogger(__name__)
+
+
+def customer_id_getter(kwargs):
+    return kwargs.get("customer_id", "test_customer")
+
+
+def aflo_dimensions_getter(kwargs):
+    return kwargs.get("aflo_dimensions", {})
+
+
+def meter_llm(
+    customer_id_getter=customer_id_getter,
+    dimensions_getter=aflo_dimensions_getter,
+    metering_client=None,
+):
+    if metering_client is None:
+        metering_client = create_ingest_client(os.environ.get("API_KEY"))
+        atexit.register(metering_client.shutdown)
+    else:
+        atexit.register(metering_client.shutdown)
+
+    def decorator(func):
+        def wrapper(*args, **kwargs):
+            # Try and grab a customer_id and dimensions from the function arguments
+            customer_id = customer_id_getter(kwargs)
+            aflo_dimensions = dimensions_getter(kwargs)
+
+            llm_response = func(*args, **kwargs)
+
+            events = process_llm_response(
+                llm_response=llm_response,
+                customer_id=customer_id,
+                aflo_dimensions=aflo_dimensions,
+            )
+            events = [event for event in events if event is not None]
+            for event in events:
+                try:
+                    logger.debug(f"Sending event: {event}")
+                    metering_client.send(event)
+                except Exception as e:
+                    logger.error(f"Failed to send event: {event} with error: {e}")
```

just double check that .send doesn't already swallow exceptions

---

send will raise an exception when calling `post`

```
    def post(self, path, payload, params=None):
        data = _gzip(payload)
        response = self.session.post(self.root_url + path, data=data, params=params)
        return self._parse(response)

    def _parse(self, response, raw=False):
        """
        Returns the raw response or raise an exception on errors.
        """
        if response.status_code != 200:
            self.logger.error("%s: %s", response.status_code, response.text)
            raise ApiError(
                response.status_code,
                response.text,
            )
        return response.text
```